core/trino-main/src/test/java/io/trino/execution/scheduler/TestUniformNodeSelector.java

```diff
@@ -0,0 +1,328 @@
+        remoteTaskScheduledExecutor = newScheduledThreadPool(2, daemonThreadsNamed("remoteTaskScheduledExecutor-%s"));
+
+        finalizerService.start();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown()
+    {
+        remoteTaskExecutor.shutdown();
+        remoteTaskExecutor = null;
+        remoteTaskScheduledExecutor.shutdown();
+        remoteTaskScheduledExecutor = null;
+        nodeSchedulerConfig = null;
+        nodeScheduler = null;
+        nodeSelector = null;
+        finalizerService.destroy();
+        finalizerService = null;
+    }
+
+    @Test
+    public void testQueueSizeAdjustmentScaleDown()
+    {
+        InternalNode node1 = new InternalNode("node1", URI.create("http://10.0.0.1:13"), NodeVersion.UNKNOWN, false);
+        nodeManager.addNodes(node1);
+        InternalNode node2 = new InternalNode("node2", URI.create("http://10.0.0.1:12"), NodeVersion.UNKNOWN, false);
+        nodeManager.addNodes(node2);
+
+        TestingTicker ticker = new TestingTicker();
+        UniformNodeSelector.QueueSizeAdjuster queueSizeAdjuster = new UniformNodeSelector.QueueSizeAdjuster(10, 100, ticker);
+
+        nodeSelector = new UniformNodeSelector(
+                nodeManager,
+                nodeTaskMap,
+                false,
+                () -> createNodeMap(TEST_CATALOG_HANDLE),
+                10,
+                100,
+                10,
+                500,
+                NodeSchedulerConfig.SplitsBalancingPolicy.STAGE,
+                false,
+                queueSizeAdjuster);
+
+        Set<Split> splits = new LinkedHashSet<>();
+
+        for (int i = 0; i < 20; i++) {
+            splits.add(new Split(TEST_CATALOG_HANDLE, TestingSplit.createRemoteSplit()));
+        }
+
+        // assign splits, mark all splits running to trigger adjustment
+        Multimap<InternalNode, Split> assignments1 = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        assertEquals(assignments1.size(), 2);
+        MockRemoteTaskFactory remoteTaskFactory = new MockRemoteTaskFactory(remoteTaskExecutor, remoteTaskScheduledExecutor);
+        int task = 0;
+        for (InternalNode node : assignments1.keySet()) {
+            TaskId taskId = new TaskId(new StageId("test", 1), task, 0);
+            task++;
+            MockRemoteTaskFactory.MockRemoteTask remoteTask = remoteTaskFactory.createTableScanTask(taskId, node, ImmutableList.copyOf(assignments1.get(node)), nodeTaskMap.createPartitionedSplitCountTracker(node, taskId));
+            remoteTask.startSplits(remoteTask.getPartitionedSplitsInfo().getCount()); // mark all task running
+            nodeTaskMap.addTask(node, remoteTask);
+            taskMap.put(node, remoteTask);
+        }
+        Set<Split> unassignedSplits = Sets.difference(splits, new HashSet<>(assignments1.values()));
+        assertEquals(unassignedSplits.size(), 18);
+
```

Comment:
nit: add a comment, "It's possible to add new assignments because split queue was upscaled"